src/Microsoft.AspNetCore.SignalR/Internal/AsyncEnumeratorAdapters.cs

```diff
@@ -0,0 +1,125 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Linq;
+using System.Reflection;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Threading.Tasks.Channels;
+
+namespace Microsoft.AspNetCore.SignalR.Internal
+{
+    // True-internal because this is a weird and tricky class to use :)
+    internal static class AsyncEnumeratorAdapters
+    {
+        private static readonly MethodInfo _boxEnumeratorMethod = typeof(AsyncEnumeratorAdapters)
+            .GetRuntimeMethods()
+            .Single(m => m.Name.Equals(nameof(BoxEnumerator)) && m.IsGenericMethod);
+
+        private static readonly MethodInfo _fromObservableMethod = typeof(AsyncEnumeratorAdapters)
+            .GetRuntimeMethods()
+            .Single(m => m.Name.Equals(nameof(FromObservable)) && m.IsGenericMethod);
+
+        private static readonly object[] _getAsyncEnumeratorArgs = new object[] { CancellationToken.None };
+
+        public static IAsyncEnumerator<object> FromObservable(object observable, Type observableInterface)
+        {
+            // TODO: Cache expressions by observable.GetType()?
+            return (IAsyncEnumerator<object>)_fromObservableMethod
+                .MakeGenericMethod(observableInterface.GetGenericArguments())
+                .Invoke(null, new[] { observable });
+        }
+
+        public static IAsyncEnumerator<object> FromObservable<T>(IObservable<T> observable)
+        {
+            // TODO: Allow bounding and optimizations?
+            var channel = Channel.CreateUnbounded<object>();
+            var cancellationTokenSource = new CancellationTokenSource();
+
+            var subscription = observable.Subscribe(new ChannelObserver<T>(channel.Out, cancellationTokenSource.Token));
```

Comment:
Because at one point I planned to cancel it, but it turns out there's nowhere to do so. Will fix :)